### PR TITLE
Move Bundler API to Rubygems.org

### PR DESCRIPTION
**Pull request by maclover7**
PR to combine Bundler API's Sinatra app with Rubygems.org's Rails app.

cc @indirect 

TODO:
- [x] Import bundler-api code
- [x] Rails-ify
- [x] Remove duplicate tasks?
- [x] Figure out subdomain/routing stuff --> use the "api/v1/dependencies" endpoint
- [x] Add caching
- [x] Provision memcached in production (@dwradcliffe)
- [ ] Make sure tests are passing
- [ ] performance/load testing
- [ ] :tada:

### CODE REVIEW

**Comment by dwradcliffe** [comment]
Let's get rid of librato before doing this.

**Comment by maclover7** [comment]
What metrics service are we moving to instead of Librato?

**Comment by segiddins** [comment]
@maclover7 maybe a subtree merge would be better, so we can preserve history?

**Comment by indirect** [comment]
@segiddins since I expect this PR to reimplement the API as a rails controller before it's merged, I think the history probably isn't worth it in this particular case. Good idea, though.

**Comment by arthurnn** [comment]
Agree with @indirect . I dont think we should port the code honestly.
We should rewrite it in a Rails controller, from scratch, having the new api as base line.
I think old bundler-api server can still run for the old api, and rubygems.org will only have the new API.

**Comment by maclover7** [comment]
What domain/sub domain will we be running the API on? bundler.rubygems.org? We have to keep in mind that v2 (new index) API will be coming out soon...

**Comment by bf4** [comment]
any reason not to do this as an engine/rack app in another repo and mount it?

**Comment by indirect** [comment]
@b4f because it would make the rg.org depend on more stacks to run (Sinatra/sequel as well as Rails/AR). This isn't super important, but it's a nice to have as we transition away from the API being the main index that people use.

**Comment by maclover7** [comment]
@indirect @qrush Does the current Bundler API and the Rubygems.org API overlap at all with their functionality? (I'm not very familiar with the codebase)

**Comment by indirect** [comment]
@maclover7 nope, no overlap. The Bundler API only provides dependencies, and the current Rubygems API has no equivalent.

**Comment by maclover7** [comment]
Gotcha. Was just wondering if everything needed to be written from scratch, or if there was any starting point. I know @dwradcliffe said we would be moving off of Librato -- what stats logger are we moving to?

**Comment by dwradcliffe** [comment]
@maclover7 We're moving everything to datadog/statsd. I think @andremedeiros is working on migrating it.

**Comment by maclover7** [comment]
Got done a lot of work with @indirect, we Rails-ified the web process. Only thing left to do is to integrate Memcached (ping @dwradcliffe).

File: app/controllers/api/v1/dependencies_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class Api::V1::DependenciesController < Api::BaseController
+  skip_before_action :verify_authenticity_token
+  GEM_REQUEST_LIMIT = 250
+
+  def index
+    return render text: "Too many gems!", status: 422 if gem_names.size > GEM_REQUEST_LIMIT
+
+    deps = gem_names.each_with_object([]) do |gem_name, deps_array|
```
Comment (qrush) [inline app/controllers/api/v1/dependencies_controller.rb]: Can we get this all in a model? `DependencyResolver` or something. This is a ton of logic for a controller.

File: app/controllers/api/v1/dependencies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+class Api::V1::DependenciesController < Api::BaseController
+  skip_before_action :verify_authenticity_token
+  GEM_REQUEST_LIMIT = 250
+
+  def index
+    return render text: "Too many gems!", status: 422 if gem_names.size > GEM_REQUEST_LIMIT
```
Comment (qrush) [inline app/controllers/api/v1/dependencies_controller.rb]: Could we link to a help page about what to do if you have too many gems?

File: app/controllers/api/v1/dependencies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+          :number => version.number,
+          :platform => version.platform,
+          :dependencies => version_deps.map { |d| [d.name, d.requirements] }
+        }
+      end
+      deps_array << gem_deps_hashes
+    end
+    respond_to do |format|
+      format.json { render json: deps }
+      format.marshal { Marshal.dump(deps) }
```
Comment (qrush) [inline app/controllers/api/v1/dependencies_controller.rb]: These are both going to be pretty expensive. What's the story for caching these requests?

File: app/controllers/api/v1/dependencies_controller.rb (hunk 1 above)
Comment (maclover7) [inline app/controllers/api/v1/dependencies_controller.rb]: So, we would call DependencyResolver.new(gem_names) ?

**Comment by qrush** [comment]
Can we port over the bundler-api tests too?

**Comment by maclover7** [comment]
Do we want to port those over from RSpec?

File: app/controllers/api/v1/dependencies_controller.rb (hunk 1 above)
Comment (qrush) [inline app/controllers/api/v1/dependencies_controller.rb]: Sounds good. And then you can `to_json` or `to_marshal` it in the `respond_to` block.

File: app/controllers/api/v1/dependencies_controller.rb (hunk 2 above)
Comment (indirect) [inline app/controllers/api/v1/dependencies_controller.rb]: There's nothing to do... Bundler automatically falls back on the full index.

**Comment by qrush** [comment]
Without a doubt. It should be pretty straightforward.

File: app/controllers/api/v1/dependencies_controller.rb (hunk 3 above)
Comment (indirect) [inline app/controllers/api/v1/dependencies_controller.rb]: As noted in this PR, we'll be adding the same memcached setup as the current API in later commits.

**Comment by maclover7** [comment]
@indirect Removed the memcached travis service commit. Rebased on master, and now Github is saying there are merge conflicts (which there aren't -- git locally on my machine says I'm up to date). Travis doesn't create a build job unless Github says there are no merge conflicts

**Comment by dwradcliffe** [comment]
@maclover7 merge conflict is probably in the routes

**Comment by segiddins** [comment]
@maclover7 any interest in revisiting this?

**Comment by maclover7** [comment]
I'll take a look at this tonight, and bring this up to speed. :)

**Comment by maclover7** [comment]
Uh, what? Rebasing appears to be have closed my PR...?

**Pull request closed without merging**